Add Supported types of tests to MUnit docs (V2.1) (#86)
mulesoft/docs-munit: modules/ROOT/pages/supported-types-of-tests.adoc

> = MUnit Supported Types of Tests

Reviewer: ```suggestion
= Types of Tests Supported by MUnit
```

ifndef::env-site,env-github[]
include::_attributes.adoc[]
endif::[]

> The following are definitions and concepts that MuleSoft recommends to take into account while designing and organizing test cases for your Mule applications. Understanding these definitions and concepts will help you build more robust MUnit tests by covering multiple edge cases, automate your tests through your CI/CD pipeline, and maximize the quality coverage for your Mule application.

Reviewer: GLOBALLY avoid introducing topics directly by referring to "the following text/section/definitions/concepts...."
Get straight to the point by summarizing the content of the topic set, addressing the audience directly as "you" and telling "you" why they want or need this information, to achieve a specific user goal:
```suggestion
Before you design and organize test cases for your Mule applications, you should understand the types of tests supported by MUnit: unit testing, integration testing, and functional testing. Having this knowledge can help you to build more robust MUnit tests by covering multiple edge cases, to automate your tests through your CI/CD pipeline, and to maximize the quality coverage for your Mule application.
```

> As a general good practice, MuleSoft recommends following a naming convention for file names and tests so that you can quickly realize what was being tested when a test fails.

Reviewer: ```suggestion
As a general good practice, be sure to follow a naming convention for file and tests that enables you to quickly identify what was being tested it a test fails.
```

When testing a Mule application MUnit allows you to cover three different types of tests:

* Unit tests.
* Integration tests.

> * Functional test.

Reviewer: OMIT in favor of introducing the three types right at the beginning, as suggested above.

== Unit Testing

> Unit testing consists of testing individual units or components of a program. A unit is the smallest testable part of any software. For a Mule application, these units are flows and when doing unit testing for flows, you must isolate them from the rest of the Mule application or external services by mocking these calls to any resource outside the flow you are testing.

Reviewer: ```suggestion
Unit testing consists of testing the smallest testable part of any software. For a Mule application, this unit is the message flow. When performing unit testing for a message flow, you must isolate it from the rest of the Mule application or external services by mocking the message to a resource outside the flow you are testing.
```

> Creating these tests helps you validate the behavior of key processors and the behavior of a flow as a whole.

Reviewer: ```suggestion
Performing unit testing helps you to validate the behavior of key processors and of the flow as a whole.
```

> == Integration testing

Reviewer: ```suggestion
== Integration Testing
```

When running integration tests you must validate how the units interact with each other or with external systems to expose faults in the interactions. Samples of integration testing in a Mule application are:

> * Testing how a service responds (Web Services, Salesforce, ServiceNow, Amazon, etc).

Reviewer: ```suggestion
* Testing how a service such as a web service, Salesforce, ServiceNow, or Amazon responds
```

Reviewer: Should we add what they respond to: "...responds to a message flow"?

> * Testing a database driver.

Reviewer: ```suggestion
* Testing a database driver
```

> * Validating a database response data structure.

Reviewer: ```suggestion
* Validating a database response data structure
```

> * Testing in-between flows interactions.

Reviewer: Does this mean testing the interactions between stages of a flow?
Are the parts of flows called "stages"?
Will the audience wonder if the smallest testable unit is the stage? Prolly not....

> Creating and executing these types of tests requires live versions of the services that your test cases will target.

Reviewer: ```suggestion
Creating and executing integration tests requires live versions of the services that your test cases can target.
```
"live" vs. what? Aren't unit tests performed against live resources, even if the flows are mocked?

> MuleSoft recommends running the test cases using environment properties and versions of the services dedicated to this end.

Reviewer: GLOBALLY reword to avoid "MuleSoft recommends" by using imperative voice ("do this") or, as necessary, "you should."

Reviewer: Does this sentence refer to the one before it?
Do we mean...
```suggestion
Because integration testing requires live targets, run your test cases using environment properties and versions of the services dedicated to testing.
```

> == Functional testing

Reviewer: ```suggestion
== Functional Testing
```

> Functional testing is a wider concept as it means testing the functional requirements of your Mule Application. The goal is to think in your Mule application use case and design test cases that target that functionality.

Reviewer: ```suggestion
Functional testing involves testing the use cases envisioned for your Mule application as a whole.
```

> For example, when testing a Mule application that is meant to keep a two-way synchronization between two systems, you can design a test case that creates an entry in each system and verifies the replication in the opposite. If you are testing a batch processing application, you can run a test with a representative set of information and validate the results of the batch. Or if you are testing a service orchestration, you can generate the input and test the result in the other end after the event went through different services.

Reviewer: ```suggestion
For example, to test a Mule application that is meant to maintain two-way synchronization between two systems, you can design a test case that creates an entry in each system and verifies the replication in the opposite. Or, if you want to test a batch-processing application, you can run a test with a representative set of information and validate the results of the batch. Finally, to test service orchestration, you can generate an event at one end of a flow and test the result at the other end, after the event went through different services.
```

> MuleSoft recommends running the test cases using environment properties and version of the services dedicated to this end.

Reviewer: ```suggestion
As with integration testing, run your functional test cases using environment properties and versions of the services dedicated to testing.
```
